src/models/model.cpp

```diff
@@ -224,20 +217,25 @@ int32_t Tokenizer::TokenToTokenId(const char* token) const {
   return token_id;
 }
 
-#if USE_CUDA
 // Since Python/Others can and will hold onto a generator object past the model object's lifetime we need to ensure
 // the allocator used is not destroyed until last. This keeps the allocator around until exit, after all other memory
 // has been destroyed. Without this, we will crash in the Onnxruntime BFCArena code when deleting tensors due to the
 // arena already being destroyed.
-Ort::Allocator* GetCudaAllocator(OrtSession& session) {
-  auto& globals = *GetOrtGlobals();
-  if (!globals.allocator_cuda_) {
-    globals.memory_info_cuda_ = OrtMemoryInfo::Create("Cuda", OrtAllocatorType::OrtDeviceAllocator, 0, OrtMemType::OrtMemTypeDefault);
-    globals.allocator_cuda_ = Ort::Allocator::Create(session, *globals.memory_info_cuda_);
+Ort::Allocator* GetDeviceAllocator(OrtSession& session, DeviceType type) {
+  // CPU Allocator is a special case, so don't try to create it
+  if (type == DeviceType::CPU)
+    return &GetDeviceInterface(DeviceType::CPU)->GetAllocator();
+
+  auto& device = GetOrtGlobals()->allocator_device_[static_cast<int>(type)];
+  if (!device) {
+    static const char* device_type_names[static_cast<int>(DeviceType::MAX)] = {"CPU - SEE ABOVE", "Cuda", "DML", "WebGPU Buffer"};
```

Comment:
what about `DeviceType::QNN`?

I think "WebGPU Buffer" should be "WebGPU_Buffer".
https://github.com/microsoft/onnxruntime/blob/e3e41739a7ca0ce0806805aa7e2814c72748d0e5/include/onnxruntime/core/framework/allocator.h#L56
https://github.com/microsoft/onnxruntime/blob/e3e41739a7ca0ce0806805aa7e2814c72748d0e5/onnxruntime/core/framework/allocator.cc#L143

Reply:
Good catch on WebGPU_Buffer.

Doesn't QNN use CPU memory? It doesn't have a device allocator "QnnWithSharedMemory".

Reply:
Actually, you gave me an idea, this should fail to compile if new providers are added. I fixed it.